lucene/core/src/java/org/apache/lucene/document/IndexSortDocValuesRangeQuery.java

```diff
@@ -0,0 +1,279 @@
+      }
+
+      @Override
+      public boolean isCacheable(LeafReaderContext ctx) {
+        return DocValues.isCacheable(ctx, field);
+      }
+    };
+  }
+
+  /**
+   * Computes the document IDs that lie within the range [lowerValue, upperValue] by
+   * performing binary search on the field's doc values.
+   *
+   * Because doc values only allow forward iteration, we need to reload the field comparator
+   * every time the binary search accesses an earlier element.
+   */
+  private static DocIdSetIterator getDocIdSetIterator(long lowerValue,
+                                                      long upperValue,
+                                                      SortField sortField,
+                                                      LeafReaderContext context,
+                                                      DocIdSetIterator delegate) throws IOException {
+    int maxDoc = context.reader().maxDoc();
+    if (maxDoc <= 0) {
+      return DocIdSetIterator.empty();
+    }
+
+    if (sortField.getReverse()) {
+      long temp = lowerValue;
+      lowerValue = upperValue;
+      upperValue = temp;
+    }
+
+    // Perform a binary search to find the first document with value >= lowerValue.
+    ValueComparator comparator = loadComparator(sortField, lowerValue, context);
+    int low = 0;
+    int high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) <= 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, lowerValue, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+    int firstDocIdInclusive = high + 1;
+
+    // Perform a binary search to find the first document with value > upperValue.
+    // Since we know that upperValue >= lowerValue, we can initialize the lower bound
+    // of the binary search to the result of the previous search.
+    comparator = loadComparator(sortField, upperValue, context);
+    low = firstDocIdInclusive;
+    high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) < 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, upperValue, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+
+    int lastDocIdExclusive = high + 1;
+    return new BoundedDocSetIdIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
+  }
+
+  /**
+   * Compares the given document's value with a stored reference value.
+   */
+  private interface ValueComparator {
+    int compare(int docID) throws IOException;
+  }
+
+  private static ValueComparator loadComparator(SortField sortField,
+                                                long topValue,
+                                                LeafReaderContext context) throws IOException {
+    @SuppressWarnings("unchecked")
+    FieldComparator<Long> fieldComparator = (FieldComparator<Long>) sortField.getComparator(1, 0);
```

Comment:
Since we only need to compare values from a single source I don't think we should use a FieldComparator here. Using the docvalues iterator directly should be possible so we don't need to add this extra indirection ?

Reply:
I used a field comparator because it dealt with missing values (otherwise I would have to handle missing values explicitly and also make sure to use the right default missing value). This approach felt a bit more solid, but I don't have a strong preference.

Reply:
Got it thanks, we need to handle missing values too so +1 for this approach.